File: Array/LC_128_Longest_Consecutive_Sequence.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.17

class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        unordered_set<int> seen(nums.begin(), nums.end());
        /*
        unordered_set<int> seen;
        for (int num: nums){
            seen.insert(num);
        }*/
        int max_length = 0;
        for (int num : nums){
            if (seen.count(num-1)){
                continue;
            }
            int curlen = 1;
            while (seen.count(num + curlen)){
                curlen += 1;
            }
            max_length = max(max_length, curlen);
        }
        
        return max_length;
    }
};

File: Array/LC_128_Longest_Consecutive_Sequence.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.17

class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        # tricky problem!
        if not nums:
            return 0
        nums_set = set(nums)
        ans = 1

        for num in nums_set:
            if num - 1 not in nums_set:
                curlen = 0
                while num in nums_set:
                    num += 1
                    curlen += 1
                ans = max(ans, curlen)
        return ans